### PR TITLE
python3Packages.tensorflow: 2.4.1 -> 2.5.0

### PR DESCRIPTION
**Pull request by Rakesh4G**
python3Packages.tensorflow: 2.4.1 -> 2.5.0

###### Motivation for this change
https://github.com/tensorflow/tensorflow/releases/tag/v2.5.0

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Rakesh4G** [comment]
I need to update commit message.

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 1]
```diff
@@ -110,7 +110,7 @@ let
       owner = "tensorflow";
       repo = "tensorflow";
       rev = "v${version}";
-      sha256 = "sha256-J62QfP45g5nxN9Nqa1tAGyc4vD2JKh50ddHLrd6/qsY=";
+      sha256 = "0v5qfzkfj9p7z50kp6zfq584n4pzyvjhsajbf5mpydgw4zm7af13";
```
Comment (veprbl) [inline pkgs/development/python-modules/tensorflow/default.nix:113]: ``
error: hash mismatch in fixed-output derivation '/nix/store/llbjfvdl6j8rsvm9zfg3qjyr08zvsjfy-source.drv':
         specified: sha256-Izh16if8NX9rcUsqDeX2/xJLUMHumztB+ecm6eZ3uGw=
            got:    sha256-lBSQYIR0gR33O606+/wlT28F6loHLqVMw98AnEMUvMk=
```

Comment (Rakesh4G) [reply pkgs/development/python-modules/tensorflow/default.nix:113]: yes, i also getting few errors currently. I had used hash using `nix-prefetch-url`

I will try to fix these.

If there is any option to mark this PR as WIP(Work in Progress)?

Comment (veprbl) [reply pkgs/development/python-modules/tensorflow/default.nix:113]: > If there is any option to mark this PR as WIP(Work in Progress)?

Look for "Still in progress? Convert to draft" link on this page.

**Comment by CMCDragonkai** [comment]
If we want TF to be able to load keras weights with the new hdf5, we need TF 2.5.

We could have this working and backport it to 21.05 to resolve this issue: https://github.com/NixOS/nixpkgs/issues/124599

**Comment by CMCDragonkai** [comment]
I've updated the sha256 hash for the source it seems to pass that problem.

But there's also the patch. The patch file for the `setup.py` is now different and I've updated that.

However I've noticed that TF2.5.0 also requires now `keras-nightly`:

```
     # Use keras-nightly at the moment.
     'keras-nightly ~= 2.5.0.dev',
```

This was not in v2.4.0, and so this is a new dependency that isn't even in Nixpkgs atm.

Also upgrading to v2.5.0 will require also updating all the binary versions too, not just the source derivation.

**Comment by CMCDragonkai** [comment]
This will need to be packaged up as well!! https://pypi.org/project/keras-nightly/#history

Can someone else who is familiar with keras set this up?

**Comment by yorickvP** [comment]
Trying this, I'm getting
```
ERROR: /build/source/WORKSPACE:23:14: no such package '@com_github_grpc_grpc//src/compiler': BUILD file not found in directory 'src/compiler' of external repository @com_github_grpc_grpc. Add a BUILD file to a directory to mark it as a package. and referenced by '//external:grpc_cpp_plugin'
```

**Comment by yorickvP** [comment]
Disabling the TF_SYSTEM_LIBS seems to help it along for now

**Comment by CMCDragonkai** [comment]
@yorickvP how did you get around keras-nightly?

**Comment by yorickvP** [comment]
@CMCDragonkai keras-nightly (and the other python dependencies) are not required during the build, as far as I can tell.

**Comment by yorickvP** [comment]
workspace.patch:
```diff
diff --git a/tensorflow/workspace3.bzl b/tensorflow/workspace3.bzl
index 16f7176d..0c0038b6 100644
--- a/tensorflow/workspace3.bzl
+++ b/tensorflow/workspace3.bzl
@@ -3,6 +3,14 @@
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
 def workspace():
+
+    http_archive(
+        name = "io_bazel_rules_docker",
+        sha256 = "1698624e878b0607052ae6131aa216d45ebb63871ec497f26c67455b34119c80",
+        strip_prefix = "rules_docker-0.15.0",
+        urls = ["https://github.com/bazelbuild/rules_docker/releases/download/v0.15.0/rules_docker-v0.15.0.tar.gz"],
+    )
+
     http_archive(
         name = "io_bazel_rules_closure",
         sha256 = "5b00383d08dd71f28503736db0500b6fb4dda47489ff5fc6bed42557c07c6ba9",
```

**Review by jonringer — COMMENTED** [review]
getting failures with the workspace patch:
```
builder for '/nix/store/rnihdl75h1hjbsb5w5lndgqq2bd3hfv7-tensorflow-2.5.0-deps.tar.gz.drv' failed with exit code 1; last 10 log lines:
  unpacking sources
  unpacking source archive /nix/store/dilp69q5sl8dp3ap9rdv846vswc9mfby-source
  source root is source
  patching sources
  applying patch /nix/store/n5axpn1z4vxs0k70rding3amddsgwmbr-relax-dependencies.patch
  patching file tensorflow/tools/pip_package/setup.py
  applying patch /nix/store/r8f344p126557b641i3l3krxwnffn5ss-workspace.patch
  patching file WORKSPACE
  Hunk #1 FAILED at 12.
  1 out of 1 hunk FAILED -- saving rejects to file WORKSPACE.rej
```

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 125328 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>12 packages marked as broken and skipped:</summary>  <ul>    <li>python38Packages.arviz</li>    <li>python38Packages.dm-sonnet</li>    <li>python38Packages.graph_nets</li>    <li>python38Packages.optuna</li>    <li>python38Packages.pymc3</li>    <li>python38Packages.rl-coach</li>    <li>python38Packages.tensorflow-probability</li>    <li>python39Packages.dm-sonnet</li>    <li>python39Packages.graph_nets</li>    <li>python39Packages.optuna</li>    <li>python39Packages.rl-coach</li>    <li>python39Packages.tensorflow-probability</li>  </ul></details><details>  <summary>17 packages failed to build:</summary>  <ul>    <li>gpt2tc</li>    <li>python38Packages.baselines</li>    <li>python38Packages.edward</li>    <li>python38Packages.mask-rcnn</li>    <li>python38Packages.scikit-tda</li>    <li>python38Packages.tensorflow (python38Packages.tensorflow-build ,python38Packages.tensorflow-build_2 ,python38Packages.tensorflowWithoutCuda ,python38Packages.tensorflow_2)</li>    <li>python38Packages.tensorflow-bin (python38Packages.tensorflow-bin_2)</li>    <li>python38Packages.tensorflowWithCuda</li>    <li>python38Packages.tflearn</li>    <li>python38Packages.umap-learn</li>    <li>python39Packages.baselines</li>    <li>python39Packages.edward</li>    <li>python39Packages.mask-rcnn</li>    <li>python39Packages.tensorflow (python39Packages.tensorflow-build ,python39Packages.tensorflow-build_2 ,python39Packages.tensorflowWithoutCuda ,python39Packages.tensorflow_2)</li>    <li>python39Packages.tensorflowWithCuda</li>    <li>python39Packages.tflearn</li>    <li>tts</li>  </ul></details>

**Comment by jonringer** [comment]
..... why did my script go forward with the merge...

**Comment by jonringer** [comment]
Sorry for accidental merge.

There's still patch failures

**Comment by CMCDragonkai** [comment]
What do we do?

**Comment by CMCDragonkai** [comment]
Patch revert and continue with this MR?

**Comment by jonringer** [comment]
I reverted the changes to master. Not sure if you can re-open a merged PR. I would just open a new one. Sorry for the inconvience.
